python/cuxfilter/assets/datetime.py

```diff
@@ -0,0 +1,123 @@
+import numpy as np
+import cupy as cp
+import pandas as pd
+import datetime
+import cudf
+import math
+
+from ..charts.constants import CUDF_DATETIME_TYPES, CUDF_TIMEDELTA_TYPE
+
+dt = {
+    CUDF_DATETIME_TYPES[0]: 9,
+    CUDF_DATETIME_TYPES[1]: 12,
+    CUDF_DATETIME_TYPES[2]: 15,
+    CUDF_DATETIME_TYPES[3]: 18,
+}
+
+dt_unit = {9: "s", 12: "ms", 15: "us", 18: "ns"}
+
+
+def date_to_int(date):
+    if type(date) == np.datetime64:
+        date = date.astype("int64")
+    return date
+
+
+def get_dt_unit_factor(date, _type):
+    if type(date) == datetime.datetime:
+        date = date.timestamp()
+
+    _pow = dt[str(_type)] - int(math.log10(date_to_int(date)))
+    return math.pow(10, _pow)
+
+
+def to_datetime(dates):
+    """
+    Description:
+        Convert dates to pd.to_datetime while also figuring out
+        date_type unit
+    -------------------------------------------
+    Input:
+        dates = list or tuple of dates
+    -------------------------------------------
+    Output:
+        pd.to_datetime object
+    """
+    unit = {}
+    test_date = dates[0] if type(dates) in [list, tuple, pd.Series] else dates
+    if type(test_date) != datetime.datetime:
+        unit["unit"] = dt_unit[int(math.log10(date_to_int(test_date)))]
+    return pd.to_datetime(dates, **unit)
+
+
+def to_dt_if_datetime(dates, _type):
+    """
+    Description:
+        convert to datetime.datetime if _type in CUDF_DATETIME_TYPES
+
+    -------------------------------------------
+    Input:
+        dates = list or tuple of integer timestamps objects
+        _type = dtype
+    -------------------------------------------
+
+    Ouput:
+        list of datetime.datetime objects
+    """
+    if _type in CUDF_DATETIME_TYPES and type(dates) in [list, tuple]:
+        return type(dates)(to_datetime(dates).to_pydatetime())
+    return dates
+
+
+def to_np_dt64_if_datetime(dates, _type):
+    """
+    Description:
+        convert to np.datetime64 if _type in CUDF_DATETIME_TYPES
+
+    -------------------------------------------
+    Input:
+        dates = list or tuple of datetime.datetime objects
+        _type = dtype
+    -------------------------------------------
+
+    Ouput:
+        list of np.datetime64 objects
+
+    """
+    if _type in CUDF_DATETIME_TYPES and type(dates) in [list, tuple]:
+        dates = to_datetime(dates)
+        return type(dates)([date.to_datetime64() for date in dates])
+    return dates
+
+
+def check_series_for_nan(ser_):
+    if ser_.dtype == "float64":
+        return ser_[~cp.isnan(ser_)].shape[0] > 0
+    return True
```

Comment:
returns True if length of series without NaN rows is greater than 0, should actually be `if series.dtype in [int, float]:`.

Reply:
`int` arrays can't have any NaN values ever, so that case would always return False unless I am misunderstanding

Edit: unless something is different for cupy?

Reply:
ah yes, you are right,  my bad. So the condition will be `return True if length of series without NaN rows is greater than 0, when dtype == float`, fixing it

Reply:
fixed

Reply:
Is it guaranteed that that there will only ever be 64 bit float dtypes?
```
In [1]: import cupy as cp                                                                                                                              

In [2]: a32 = cp.array([1,2,3], dtype=cp.float32)                                                                                                      

In [3]: a64 = cp.array([1,2,3], dtype=cp.float64)                                                                                                      

In [4]: a64.dtype == float                                                                                                                             
Out[4]: True

In [5]: a32.dtype == float                                                                                                                             
Out[5]: False
```
If this needs to work on 32bit arrays I think it needs to use `issubdtype` to be more more general:
```
In [12]: np.issubdtype(a32.dtype, np.floating)                                                                                                         
Out[12]: True

In [13]: np.issubdtype(a64.dtype, np.floating)                                                                                                         
Out[13]: True
```

Also maybe there is a better function name? "check series for nan" sounds exactly like "return true if there is a nan in the series", but it's almost the opposite of that? What about `check_no_nan_runs`? That says (to me) "return true if there is not a run of nans"

Reply:
actually is this just checking that the array is not *all* nans?
```
In [14]: not all(np.isnan(a32))                                                                                                                        
Out[14]: True

In [15]: not all(np.isnan(a64))                                                                                                                        
Out[15]: True

In [16]: ai = cp.array([1,2,3], dtype=cp.int32)                                                                                                        

In [17]: not all(np.isnan(ai))                                                                                                                         
Out[17]: True

```
?